### PR TITLE
fix(core,schemas): use database index to prevent subdomain conflict

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Create an unique index for protected app's free subdomain, to ensure that domains will not conflict.

The previous solution won't work because of RLS.

## Testing

Local tested and unit tests.

截屏2024-01-29 下午2 31 25

## Checklist

- [ ] `.changeset`
- [x] unit tests
- [ ] integration tests
- [ ] necessary TSDoc comments

### CODE REVIEW

**Review by darcyYe — APPROVED** [review]

File: packages/core/src/libraries/protected-app.ts [hunk 1]
```diff
@@ -61,6 +61,46 @@ const getDefaultDomain = async () => {
...
+  assertThat(
+    isValidSubdomain(subDomain),
+    new RequestError({
+      code: 'application.invalid_subdomain',
+      status: 422,
+    })
+  );
+
+  // Skip for integration test, use empty value instead
+  const { domain } = EnvSet.values.isIntegrationTest ? { domain: '' } : await getProviderConfig();
```
Comment (simeng-li) [inline packages/core/src/libraries/protected-app.ts:87]: Feels like there might be multiple places calling getProviderConfig. should we place the `isIntegrationTest`  check within this method?

File: packages/core/src/routes/applications/application.ts [hunk 2]
```diff
@@ -166,29 +168,42 @@ export default function applicationRoutes<T extends AuthedRouter>(
...
+      } catch (error: unknown) {
+        if (
+          error instanceof UniqueIntegrityConstraintViolationError &&
+          error.constraint === protectedAppMetadataHostConstraint
+        ) {
+          throw new RequestError({
+            code: 'application.protected_application_subdomain_exists',
+            status: 422,
+          });
+        }
```
Comment (simeng-li) [inline packages/core/src/routes/applications/application.ts:204]: can move this to the koaErrorHandler level?

File: packages/schemas/tables/applications.sql [hunk 3]
```diff
@@ -23,3 +23,8 @@ create index applications__id
 
 create index applications__is_third_party
   on applications (tenant_id, is_third_party);
+
+create unique index applications__protected_app_metadata_host
```
Comment (simeng-li) [inline packages/schemas/tables/applications.sql:27]: just curious what's the difference between a unique index and a unique constraint?

Comment (wangsijie) [reply packages/schemas/tables/applications.sql:27]: Index will improve performance (and use more space)

File: packages/core/src/routes/applications/application.ts (hunk 2 above)
Comment (wangsijie) [reply packages/core/src/routes/applications/application.ts]: updated

File: packages/core/src/libraries/protected-app.ts (hunk 1 above)
Comment (wangsijie) [reply packages/core/src/libraries/protected-app.ts:87]: Wait until necessary?

Comment (simeng-li) [reply packages/core/src/libraries/protected-app.ts:87]: fine to me if this is the only place

**Review by simeng-li — APPROVED** [review]

**Pull request merged**